File: me_daily/2025/10/1024/cf1061div2/maximumGCDonwhiteboard.cpp
#include <bits/stdc++.h>
using namespace std;

#define ALL(v) (v).begin(), (v).end()
#define For(i, n) for (int i = 0; i < (n); ++i)
#define FOR(i, a, b) for (int i = (a); i < (b); ++i)
#define pb push_back
#define fi first
#define se second
#define V vector
#define SORT(v) sort(ALL(v))

using umpii = unordered_map<int, int>;
using ll = long long;
using pii = pair<int,int>;
using vi = vector<int>;
using vll = vector<ll>;

const int inf = 0x3f3f3f3f;
const ll infl = 0x3f3f3f3f3f3f3f3fll;
const int mod = 1e9 + 7;

int fastio_init = [](){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 0;
}();

int main() {
    int t;
    cin >> t;
    For(_t, t) {
        int n, k;
        cin >> n >> k;
        vi a(n);
        For(i, n) cin >> a[i];
        int maxn = *max_element(ALL(a));
        vector<int> freq(n+1, 0);
        For(i, n) {
            freq[a[i]]++;
        }
        vector<int> prefix(n+1, 0);
        FOR(i, 1, n+1) {
            prefix[i] = prefix[i-1] + freq[i];
        }
        int ans = 1;
        for (int d = n; d >= 1; --d) {
            ll up = min((ll)n, 4LL * d - 1);
            int num_small = prefix[(int)up];
            int num_small_good = 0;
            for (int m = 1; m <= 3; ++m) {
                ll val = (ll)m * d;
                if (val > n) break;
                num_small_good += freq[(int)val];
            }
            int cost = num_small - num_small_good;
            if (cost <= k) {
                ans = d;
                break;
            }
        }
        cout << ans << "\n";
    }
    return 0;
}

File: me_daily/2025/10/1024/cf1061div2/maximumGCDonWhiteboard.py
from collections import defaultdict
from typing import List
import sys
input = lambda: sys.stdin.readline().strip()
II = lambda: int(input())
LII = lambda: list(map(int, input().split()))
SI = lambda: input()
LSI = lambda: input().split()

t = II()
for _ in range(t):
    n, k = LII()
    a = LII()
    freq = [0] * (n + 1)
    for x in a:
        freq[x] += 1
    prefix = [0] * (n + 1)
    for i in range(1, n + 1):
        prefix[i] = prefix[i - 1] + freq[i]
    ans = 1
    for d in range(n, 0, -1):
        up = min(n, 4 * d - 1)
        num_small = prefix[up]
        num_small_good = 0
        for m in range(1, 4):
            val = m * d
            if val > n:
                break
            num_small_good += freq[val]
        cost = num_small - num_small_good
        if cost <= k:
            ans = d
            break
    print(ans)